### PR TITLE
No more fs.readFileSync and different hash functions

### LINKED ISSUE #48: Make compatible with browsers
Seems like we're dependent on `fs` to even include this module (See https://github.com/ipfs/js-ipfs-merkle-dag/blob/893aff1940a9281e31259affed05d983ca738478/src/dag-node.js#L12 ) which needs to be fixed, seems like js-ipfs-api is broken because of this as well

### PR DESCRIPTION
**Pull request by dignifiedquire**
- `feat(dag-node): allow for different hash functions`
- `fix(dag-node: remove dependency on fs.readFileSync`

Closes #48, #49

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -13,12 +13,35 @@
...
+  + [DAGLink Class](#daglink-class)
+  + [DAGService](#dagservice)
+    - [`put`](#put)
+    - [`putStream`](#putstream)
+    - [`get`](#get)
+    - [`getStream`](#getstream)
+    - [`getRecursive`](#getrecursive)
+    - [`getRecursiveStream`](#getrecursivestream)
+    - [`remove`](#remove)
+* [License](#license)
```
Comment (daviddias) [inline README.md:44]: sweeet! Thank you 👍 

📜 for the 📜 god

File: README.md [hunk 2]
```diff
@@ -99,10 +122,12 @@ var node = new ipfsMDAG.DAGNode([<data>, <[links]>])
 
 > (property) an array of `DAGLink`s belonging to the node
 
-#### `multihash`
+#### `multihash(fn)`
```
Comment (daviddias) [inline README.md:125]: This doesn't expect an actual function, just the hash identifier. Can we use a better name for this arg?

Also, if there is a default. It should be `[<fn>]`

File: src/dag-node.js [hunk 3]
```diff
@@ -2,14 +2,12 @@
...
 const stable = require('stable')
-const fs = require('fs')
-const path = require('path')
 const mh = require('multihashes')
 
 const util = require('./util')
 const DAGLink = require('./dag-link')
 
-const proto = protobuf(fs.readFileSync(path.join(__dirname, 'merkledag.proto')))
+const proto = protobuf(require('./merkledag.proto'))
```
Comment (daviddias) [inline src/dag-node.js:10]: This is new.

File: src/dag-node.js [hunk 4]
```diff
@@ -157,12 +155,17 @@ module.exports = class DAGNode {
 
   // Encoded returns the encoded raw data version of a Node instance.
   // It may use a cached encoded version, unless the force flag is given.
-  encoded (force) {
+  encoded (fn, force) {
```
Comment (daviddias) [inline src/dag-node.js:158]: This is breaking API, `(force, fn)` wouldn't.

File: src/merkledag.proto.js [hunk 5]
```diff
@@ -1,3 +1,6 @@
+'use strict'
+
+module.exports = new Buffer(`
```
Comment (daviddias) [inline src/merkledag.proto.js:3]: this is 'ugly'.. Now I see how require worked.. this doesn't remove complexity, it even adds.

Comment (dignifiedquire) [reply src/merkledag.proto.js:3]: It removes a lot of complexity for users, not using aegir. These proto files are the reason we need to use brfs as a preprocessor everywhere. If we drop it the code works out of the box in webpack and browserify.

File: src/dag-node.js (hunk 4 above)
Comment (dignifiedquire) [reply src/dag-node.js:158]: okay will change

File: src/merkledag.proto.js (hunk 5 above)
Comment (daviddias) [reply src/merkledag.proto.js:3]: Is this the only `fs` that required to be brfs'ified in js-ipfs-api? brfs'ified is a standard process in transpiling code to the browser anyway. This seems to me more of a way to postpone getting proper documentation that explains how to do this. Which we have to write anyway.

Comment (dignifiedquire) [reply src/merkledag.proto.js:3]: The only reason for using brfs for end users today is our use of `fs.readFileSync` to load proto files. We can avoid this by simply using a pattern like this. This means reduced build time for us and our users, in addition to having one less thing developers have to worry about. `brfs` is nice, but our end goal should be that `webpack` and `browserify` work on our modules without any additional configuration. 

This is the first step for this :)

Comment (daviddias) [reply src/merkledag.proto.js:3]: @dignifiedquire my main point is: **we need to have a section in our main projects (js-ipfs, js-ipfs-api and js-libp2p) that explains our users how to 'webpack' all of this stuff**, simply, it is better to have a walkthrough, even if it is obvious than having nothing. As an example, I don't believe that js-ipfs will ever stop needing a special webpack config.

Comment (dignifiedquire) [reply src/merkledag.proto.js:3]: > we need to have a section in our main projects (js-ipfs, js-ipfs-api and js-libp2p) that explains our users how to 'webpack' all of this stuff, 

Yes, but we still should aim to make that explanation as simple as possible :)

> I don't believe that js-ipfs will ever stop needing a special webpack config.

I have concrete plans for this and a roadmap in my head on how to get there, it is coming soon. Dropping `fs.readFile` is the first step on the road there.

**Comment by dignifiedquire** [comment]
Closing as this module is deprecated now. Ref #62

**Pull request closed without merging**